SharkCage/SharedFunctionality/ValidateBinary.cpp

```diff
@@ -0,0 +1,226 @@
+	BCRYPT_HASH_HANDLE hash_handle = nullptr;
+	NTSTATUS status = STATUS_UNSUCCESSFUL;
+	DWORD data = 0, cb_hash = 0, cb_hash_object = 0;
+	PBYTE hash_object = nullptr;
+	PUCHAR hash = nullptr;
+
+	//open an algorithm handle
+	if (!NT_SUCCESS(status = ::BCryptOpenAlgorithmProvider(
+		&algorithm,
+		BCRYPT_SHA512_ALGORITHM,
+		nullptr,
+		0)))
+	{
+		CleanupHash(algorithm, hash_handle, hash_object, hash);
+		return false;
+	}
+
+	//calculate the size of the buffer to hold the hash object
+	if (!NT_SUCCESS(status = ::BCryptGetProperty(
+		algorithm,
+		BCRYPT_OBJECT_LENGTH,
+		reinterpret_cast<PBYTE>(&cb_hash_object),
+		sizeof(DWORD),
+		&data,
+		0)))
+	{
+		CleanupHash(algorithm, hash_handle, hash_object, hash);
+		return false;
+	}
+
+	//allocate the hash object on the heap
+	hash_object = reinterpret_cast<PBYTE>(::HeapAlloc(::GetProcessHeap(), 0, cb_hash_object));
+	if (!hash_object)
+	{
+		CleanupHash(algorithm, hash_handle, hash_object, hash);
+		return false;
+	}
+
+	//calculate the length of the hash
+	if (!NT_SUCCESS(status = ::BCryptGetProperty(
+		algorithm,
+		BCRYPT_HASH_LENGTH,
+		reinterpret_cast<PBYTE>(&cb_hash),
+		sizeof(DWORD),
+		&data,
+		0)))
+	{
+		CleanupHash(algorithm, hash_handle, hash_object, hash);
+		return false;
+	}
+
+	//allocate the hash buffer on the heap
+	hash = reinterpret_cast<PBYTE>(::HeapAlloc(::GetProcessHeap(), 0, cb_hash));
+	if (!hash)
+	{
+		CleanupHash(algorithm, hash_handle, hash_object, hash);
+		return false;
+	}
+
+	//create a hash
+	if (!NT_SUCCESS(status = ::BCryptCreateHash(
+		algorithm,
+		&hash_handle,
+		hash_object,
+		cb_hash_object,
+		0,
+		0,
+		0)))
+	{
+		CleanupHash(algorithm, hash_handle, hash_object, hash);
+		return false;
+	}
+
+	//hash data
+	if (!NT_SUCCESS(status = ::BCryptHashData(
+		hash_handle,
+		reinterpret_cast<PUCHAR>(buffer.data()),
+		buffer.size(),
+		0)))
+	{
+		CleanupHash(algorithm, hash_handle, hash_object, hash);
+		return false;
+	}
+
+	//close the hash
+	if (!NT_SUCCESS(status = ::BCryptFinishHash(
+		hash_handle,
+		hash,
+		cb_hash,
+		0)))
+	{
+		CleanupHash(algorithm, hash_handle, hash_object, hash);
+		return false;
+	}
+
+	std::vector<char> current_hash = BytesToHexString(hash, cb_hash);
+
+	CleanupHash(algorithm, hash_handle, hash_object, hash);
+
+	return CompareHashes(app_hash, current_hash);
+}
+
+std::vector<char> ValidateBinary::BytesToHexString(unsigned char const *bytes, std::size_t length)
+{
+	std::vector<char> file_hash;
+
+	static char const lookup[] = "0123456789ABCDEF";
+
+	for (std::size_t i = 0; i != length; ++i)
+	{
+		file_hash.push_back(lookup[bytes[i] >> 4]);
+		file_hash.push_back(lookup[bytes[i] & 0xF]);
+	}
+
+	return file_hash;
+}
+
+bool ValidateBinary::CompareHashes(const std::wstring &hash_1, const std::vector<char> &hash_2)
+{
+	std::string str_hash_2(hash_2.begin(), hash_2.end());
+
+	if (hash_1.size() < str_hash_2.size())
+	{
+		return false;
+	}
+
+	return std::equal(str_hash_2.begin(), str_hash_2.end(), hash_1.begin());
```

Comment:
if you specify the last element of `hash_1` here this should already perform a size check. from the doc: `If the length of the range [first1, last1) does not equal the length of the range [first2, last2), returns false`. So maybe just add `hash_1.last()` as the fourth parameter and delete the manual size check above